### ISSUE TITLE
invalid cni config file

### ISSUE THREAD

**Issue by little-dude** [intel/multus-cni#146]
Hi,

I'm making my first steps with multus, so this is probably a dumb mistake from me. Also, I'm not running multus as a daemonset.

I have the following cni config file:

```json
[root@k8s-master ~]# cat /etc/cni/net.d/70-multus.conflist 
{
    "name": "multus-cni-network",
    "type": "multus",
    "delegates": [
        {
            "type": "flannel",
            "name": "flannel.1",
            "delegate": {
                "isDefaultGateway": true
            }
        }
    ],
    "kubeconfig": "/etc/kube-multus/kube.config.yml"
}
```
My flannel network is defined as with the following yaml

```yaml
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: flannel-net
spec:
  config: '{
    "cniVersion": "0.3.0",
    "type": "flannel",
    "delegate": {
        "hairpinMode": true,
        "isDefaultGateway": true
    },
    {
        "type": "portmap",
        "capabilities": {
            "portMappings": true
        }
    }
  }'
```

Here is the description:

```
[root@k8s-master ~]#  kubectl get nodes k8s-master.mvcalab.us.alcatel-lucent.com
NAME                                       STATUS     ROLES     AGE       VERSION
k8s-master.mvcalab.us.alcatel-lucent.com   NotReady   <none>    22d       v1.11.2
[root@k8s-master ~]# kubectl describe net-attach-def
Name:         flannel-net
Namespace:    default
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"k8s.cni.cncf.io/v1","kind":"NetworkAttachmentDefinition","metadata":{"annotations":{},"name":"flannel-net","namespace":"default"},"spec"...                    
API Version:  k8s.cni.cncf.io/v1
Kind:         NetworkAttachmentDefinition
Metadata:
  Creation Timestamp:  2018-09-04T20:44:31Z
  Generation:          1
  Resource Version:    2994650
  Self Link:           /apis/k8s.cni.cncf.io/v1/namespaces/default/network-attachment-definitions/flannel-net
  UID:                 52d0df4f-b083-11e8-b2cc-ac1f6b034124
Spec:
  Config:  { "cniVersion": "0.3.0", "type": "flannel", "delegate": { "hairpinMode": true, "isDefaultGateway": true }, { "type": "portmap", "capabilities": { "portMappings": true } } }                                                      
Events:    <none>
```

But in the kubelet logs I see:

```
Sep 04 14:27:18 k8s-slave-1.mvcalab.us.alcatel-lucent.com kubelet[2524]: W0904 14:27:18.510329    2524 cni.go:120] Error loading CNI config list file /etc/cni/net.d/70-multus.conflist: error parsing configuration list: no 'plugins' key
Sep 04 14:27:18 k8s-slave-1.mvcalab.us.alcatel-lucent.com kubelet[2524]: W0904 14:27:18.510361    2524 cni.go:172] Unable to update cni config: No valid networks found in /etc/cni/net.d
```

**Comment by dougbtv** [comment]
Hey @little-dude -- couple things to try...

1. Seeing it's not actually a list try changing the extension to `.conf` I didn't check the code, but, I think that's how it might differentiate in how it parses it.

2. The capabilities feature is not yet pulled in -- https://github.com/intel/multus-cni/pull/132 -- so that might cause you some grief too.

Any reason why you can't use the daemonset-style install? Maybe you want to change some stuff, but... It's the most reliable way to get a baseline install.

Lastly you can also turn on the enhanced logging to get some more clues, too, more info here @ https://github.com/intel/multus-cni#logging-options

**Comment by little-dude** [comment]
Hi @dougbtv,

Thank you for the fast answer!
You're right: the file must be a `.conf` instead of `.conflist`. It seems to work after that (I mean, at least my nodes are Ready, and the pods come up, I haven't checked in detail yet).

All right for the capabilities. I'll monitor https://github.com/intel/multus-cni/pull/132

> Any reason why you can't use the daemonset-style install? Maybe you want to change some stuff, but... It's the most reliable way to get a baseline install.

No good reason I'm afraid. But when I started with kubernetes a few weeks ago, it was easier for me to use what I already knew, ie good old systemd services. As a result, in my cluster, the api-server, controller, kubelet, kube-proxy, and flannel are running as services. I realize it may be frowned upon in the k8s world though, and I'll probably do a v2 when I'm more comfortable with kubernetes.

Thanks for the tip for logging, I just added a line to have logs into a log file. Can I write log in journald already? If not I may contribute a patch :)


Also I believe your the author of this [very nice blog post](http://dougbtv.com/nfvpe/2017/12/19/multus-crd/), thank you for that to!
